### ISSUE TITLE
Select.asTable() should delay slow call to Tools::autoAlias

### ISSUE THREAD

**Issue by lukaseder** [jOOQ/jOOQ#13349]
As discussed in https://github.com/jOOQ/jOOQ/issues/6714, calling `Select.asTable()` without explicit alias can be relatively slow depending on how frequently it is called. It should be avoided from within jOOQ's internals.

One special case is `DSL.table(Select)`, which is a synonym for `select.asTable()`, although users may choose to call this method only to call `as(alias)` right after:

```java
table(selectOne()).as("t", "c")
```

In this case, it's really regrettable we generate an intermediary alias so slowly, only to throw it away again. It would be better for `DSL.table(Select)` to return a temporary data structure that materialises the alias only if any method other than `asTable(alias)` or `as(alias)` is called on the table.

Should #6714 be fixed, this intermediary data structure might be removed again, but until then, it could prevent a few "accidents."

**Comment by lukaseder** [comment]
The implementation of such a `LazyDerivedTable` couldn't inherit from `AbstractTable`, as it would have to override all methods to make sure the lazy behaviour is correct (e.g. `getName()` mustn't return any constant name, but trigger the materialization of the backing derived table, instead).

That's a bit excessive. Perhaps there's another way, or we'll just fix #6714 instead, as this improvement here also yields a bit of regression risk.

**Comment by lukaseder** [comment]
Regrettably, using `LazyName` doesn't suffice. `TableAlias` eagerly initialises its fields:

```
Thread [main] (Suspended (breakpoint at line 3397 in Tools))	
	Tools.autoAlias(QueryPart) line: 3397	
	SelectQueryImpl<R>.lambda$4() line: 654	
	0x00000008005818a8.get() line: not available [local variables unavailable]	
	LazyName.name() line: 66	
	LazyName.unqualifiedName() line: 112	
	TableAlias<R>(AbstractNamed).getUnqualifiedName() line: 81	
	TableAlias<R>.getQualifiedName() line: 165	
	AbstractNamed.qualify(Named, Name) line: 134	
	TableFieldImpl<R,T>.<init>(Name, DataType<T>, Table<R>, Comment, Binding<?,T>) line: 80	
	TableAlias<R>.lambda$2(Name[], Field, int) line: 89	
	0x00000008004ff360.apply(Object, int) line: not available	
	Tools.map(T[], ThrowingObjIntFunction<? super T,? extends U,E>) line: 2154	
	TableAlias<R>.init(Name[]) line: 89	
	TableAlias<R>.<init>(Table<R>, Name, Name[], Predicate<Context<?>>) line: 81	
	TableAlias<R>.<init>(Table<R>, Name, Predicate<Context<?>>) line: 70	
	DerivedTable<R>.as(Name) line: 77	
	SelectQueryImpl<R>.asTable() line: 654	
	SelectImpl<R,T1,T2,T3,T4,T5,T6,T7,T8,T9,T10,T11,T12,T13,T14,T15,T16,T17,T18,T19,T20,T21,T22>.asTable() line: 2861	
	AliasTests<A,B,S,B2S,BS>.testAliasingDerivedTablesWithTableType() line: 381	
	H2Test(jOOQAbstractTest<A,B,S,B2S,BS>).testAliasingDerivedTablesWithTableType() line: 6978	
	NativeMethodAccessorImpl.invoke0(Method, Object, Object[]) line: not available [native method]	
	NativeMethodAccessorImpl.invoke(Object, Object[]) line: 77	
	DelegatingMethodAccessorImpl.invoke(Object, Object[]) line: 43	
	Method.invoke(Object, Object...) line: 568	
	FrameworkMethod$1.runReflectiveCall() line: 59	
	FrameworkMethod$1(ReflectiveCallable).run() line: 12	
	FrameworkMethod.invokeExplosively(Object, Object...) line: 56	
	InvokeMethod.evaluate() line: 17	
	RunBefores.evaluate() line: 26	
	RunAfters.evaluate() line: 27	
	TestWatcher$1.evaluate() line: 61	
	TestWatcher$1.evaluate() line: 61	
	ParentRunner$3.evaluate() line: 306	
	BlockJUnit4ClassRunner$1.evaluate() line: 100	
	JUnit4(ParentRunner<T>).runLeaf(Statement, Description, RunNotifier) line: 366	
	JUnit4(BlockJUnit4ClassRunner).runChild(FrameworkMethod, RunNotifier) line: 103	
	JUnit4(BlockJUnit4ClassRunner).runChild(Object, RunNotifier) line: 63	
	ParentRunner$4.run() line: 331	
	ParentRunner$1.schedule(Runnable) line: 79	
	JUnit4(ParentRunner<T>).runChildren(RunNotifier) line: 329	
	ParentRunner<T>.access$100(ParentRunner, RunNotifier) line: 66	
	ParentRunner$2.evaluate() line: 293	
	RunBefores.evaluate() line: 26	
	RunAfters.evaluate() line: 27	
	ParentRunner$3.evaluate() line: 306	
	JUnit4(ParentRunner<T>).run(RunNotifier) line: 413	
	JUnit4TestReference.run(TestExecution) line: 93	
	TestExecution.run(ITestReference[]) line: 40	
	RemoteTestRunner.runTests(String[], String, TestExecution) line: 529	
	RemoteTestRunner.runTests(TestExecution) line: 756	
	RemoteTestRunner.run() line: 452	
	RemoteTestRunner.main(String[]) line: 210	
```

So, let's investigate lazily initialising that as well, because the `TableAlias.aliasedFields` field is just a cache. The underlying information is in `TableAlias.alias`, and its computation can be delayed as well.

**Comment by lukaseder** [comment]
Yep, that fixes it

**Comment by lukaseder** [comment]
Fixed in jOOQ 3.17.0, 3.16.6 (https://github.com/jOOQ/jOOQ/issues/13376), and 3.15.10 (https://github.com/jOOQ/jOOQ/issues/13377)
